### PR TITLE
fix: server 400 error

### PR DESCRIPTION
**Pull request by baby230211**
- add HOST_NAME to resolve current server fetch 404 error ([ALLOW_HOST](https://github.com/pycontw/pycon.tw/blob/03d3e8af584d4b3d3bcbdd23fcf05d6da70f93de/src/pycontw2016/settings/production/base.py#L19) is configed in BE)
  - add `staging.pycon.tw` in staging machine.
  - add `tw.pycon.org` in prd machine.

## Types of changes

* **Refactoring**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,12 @@
-FROM node:14.21.3-buster-slim
+FROM node:16.20.2-alpine
 
 WORKDIR /usr/local
 
-COPY package-lock.json package-lock.json
+COPY yarn.lock yarn.lock
```
Comment (mattwang44) [inline Dockerfile:5]: Any specific reason to replace npm with yarn?

File: Dockerfile [hunk 2]
```diff
@@ -1,11 +1,12 @@
-FROM node:14.21.3-buster-slim
```
Comment (mattwang44) [inline Dockerfile:1]: Any specific reason to replace debian image with alpine?

File: plugins/http.js [hunk 3]
```diff
@@ -1,6 +1,9 @@
 export default function ({ $http }) {
     const token = process.env.AUTH_TOKEN || ''
     $http.setToken(token, 'Token')
+    if (process.env.DEPLOY_ENV) {
```
Comment (mattwang44) [inline plugins/http.js:4]: How does this resolve 404 issue? What's the root cause and why does backend need it?

Comment (baby230211) [reply plugins/http.js:4]: Currently, BE specify [ALLOW_HOST](https://github.com/pycontw/pycon.tw/blob/03d3e8af584d4b3d3bcbdd23fcf05d6da70f93de/src/pycontw2016/settings/production/base.py#L19) here.
That means that we can only fetch the api with specific host header, or we will encounter 400 issue.

File: Dockerfile (hunk 2 above)
Comment (baby230211) [reply Dockerfile:1]: Just for smaller image size.

File: Dockerfile (hunk 1 above)
Comment (baby230211) [reply Dockerfile:5]: For faster install speed :).

File: Dockerfile (hunk 2 above)
Comment (baby230211) [reply Dockerfile:1]: let me fallback to buster-slim

File: plugins/http.js (hunk 3 above)
Comment (mattwang44) [reply plugins/http.js:4]: then why the value of the header is `process.env.DEPLOY_ENV` rather than the host name?

File: Dockerfile (hunk 1 above)
Comment (mattwang44) [reply Dockerfile:5]: pnpm is considered faster, more memory-efficient, and more performant than yarn, so why not use pnpm instead?

When trying to introduce/replace/ditch specific tooling, it's essential to 
- provide supplementary docs or materials to explain why this change is necessary.
- undergo impact analysis to understand the potential benefits and drawbacks.
- evaluate if there are other alternatives and make sure it's the right decision

To sum up, I believe using yarn instead would some what bring us some improvement, but it's not a good practice to just introduce it out of nowhere.

File: Dockerfile (hunk 2 above)
Comment (mattwang44) [reply Dockerfile:1]: alpine image is notorious for its DNS issue ([ref](https://martinheinz.dev/blog/92)), lack of build toolchain ([ref](https://dev.to/kakisoft/dockeralpine-why-you-should-avoid-alpine-linux-44he)), and performance issue ([ref](https://medium.com/@kruchkov.alexandr/stop-using-alpine-images-b51d12b0fde2)). If this change doesn't aim to resolve any specific issue, I'd prefer to continue using slim Debian images.

File: plugins/http.js (hunk 3 above)
Comment (mattwang44) [reply plugins/http.js:4]: How should DEPLOY_ENV be provided? It's not a nuxt/vue built-in env var, right?

Comment (baby230211) [reply plugins/http.js:4]: The naming might confuse you.
The better naming env might be `process.env.HOST_NAME`.

Comment (baby230211) [reply plugins/http.js:4]: > How should DEPLOY_ENV be provided? It's not a nuxt/vue built-in env var, right?

Yes, we should add this into the machine .env, just like we did in `AUTH_TOKEN`

File: Dockerfile (hunk 1 above)
Comment (mattwang44) [reply Dockerfile:5]: > if we use npm install, it will encounter ERESOLVE overriding peer dependency

how and when this issue be introduced? it doesn't convince me that using yarn instead would just resolve the issue.

Comment (baby230211) [reply Dockerfile:5]: pnpm sounds great. But for node version 16.
i don't think node version 16 has a great compatibility to it [ref](https://pnpm.io/installation#compatibility).
Maybe it's better for node version 18.

File: plugins/http.js (hunk 3 above)
Comment (mattwang44) [reply plugins/http.js:4]: ok, maybe rename and add it to `.env.sample` (and maybe add code comment or explain it somewhere in case we forget why we need this)

Comment (baby230211) [reply plugins/http.js]: fixed.

File: Dockerfile (hunk 2 above)
Comment (baby230211) [reply Dockerfile:1]: fixed. Thanks

File: Dockerfile (hunk 1 above)
Comment (baby230211) [reply Dockerfile]: fixed. Thanks

Comment (mattwang44) [reply Dockerfile]: it's still a good try though, we can start a discussion and have another PR for this 👍🏽

File: Dockerfile
Comment (mattwang44) [inline Dockerfile:1]: and the question now would be 
why we need to upgrade node to resolve the host name issue?

Comment (baby230211) [reply Dockerfile:1]: reverted. Thanks

**Review by mattwang44 — APPROVED** [review]
LGTM

**Review by SivanYeh — APPROVED** [review]

**Comment by SivanYeh** [comment]
@baby230211 @mattwang44 剛剛把#537合併後serve err回來了. 我猜關鍵可能藏在chore/upgrade-node在staging與repo的差異. 
不過註冊的售票頁面目前比較趕, 暫先不動staging, 等售票上線.

**Comment by mattwang44** [comment]
@baby230211 
應該是 backend ALLOWED_HOSTS 要加上 `pycontw-2024`，frontend container name 是有被拿來當 host name 塞進 header 的
image

**Comment by baby230211** [comment]
我剛剛改staging 裡的.env host name 重新deploy 就正常了。
只是如果deploy 到 production的話，需要再設定一次.env host name.。

@mattwang44 
本來我也認為需要在BE的config 裡面加上這個host，但是這個host需要每一年換一個，明年可能就會忘了。我就想說不如FE帶header進去。
另外我看FE沒有改什麼api的東西，也不確定之前在server fetch的時候有沒有噴error (說不定之前沒有在server fetch 資料)
